flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSchemaConverter.java

```diff
@@ -352,17 +457,55 @@ public static Schema convertToSchema(LogicalType logicalType, String rowName) {
                 final TimestampType timestampType = (TimestampType) logicalType;
                 precision = timestampType.getPrecision();
                 org.apache.avro.LogicalType avroLogicalType;
-                if (precision <= 3) {
-                    avroLogicalType = LogicalTypes.timestampMillis();
+                if (legacyMapping) {
+                    if (precision <= 3) {
+                        avroLogicalType = LogicalTypes.timestampMillis();
+                    } else if (precision <= 6) {
+                        avroLogicalType = LogicalTypes.timestampMicros();
+                    } else {
```

Comment:
I'm happy that you can  add precision support here, but I suggest we can open a new PR or commit to finish this work, it will be more clear.